### PR TITLE
Reconcile rather than replace inventory definitions in the store with those read from a file

### PR DESCRIPTION
**Pull request by MikeCJohnsonUk**
Currently, whenever a new set of inventory definitions is read from file, the existing store copy is completely deleted and then the new version from the file is written to the store. This generates a large number of changes which will lead to a large number of Watch events being sent to any registered listeners. This is quite taxing on the system and could well lead to short term workload failures event for minor changes to an inventory.

This edit instead compares the incoming definitions from an inventory.yaml files and compare those definitions with the set already present in the store and will only update the store where there are differences as a result of additions, updates or deletions. For cases where there are just minor changes, this will not lead to large number of Watch events and so avoid major disturbances in the simulation.

In all cases, the store copy of the definitions will match those in the incoming file once the reconciliation process is complete.

Other changes:
- ensured all testdata inventory files have names in the definition which match their filenames and purpose
- broke out longer running inventory tests into a separate module to avoid issues with test passes under VsCode hitting a test timeout limit
- small number of changes in the tracing pattern in the store layer. If this proves successful, expect further changes along the same lines.
- fixed issue with reading specific testdata inventory yaml files in face of Viper tracking originally specified file paths and names by using a separate viper instance when reading inventory definition files.

FYI: I've recently figured out how to get the editor to trim trailing white space on a line, hence lot of trimmed lines in PR. Sorry.

### CODE REVIEW

File: simulation/internal/clients/inventory/inventory.go [hunk 1]
```diff
@@ -552,23 +548,246 @@ func (m *Inventory) DeleteInventoryDefinition(ctx context.Context) error {
...
+
+		bladeStore.SetDetails(bladeFile.GetDetails())
+		bladeStore.SetCapacity(bladeFile.GetCapacity())
+		bladeStore.SetBootInfo(bladeFile.GetBootInfo())
+		bladeStore.SetBootPowerOn(bladeFile.GetBootOnPowerOn())
+
+		_, err = bladeStore.Create(ctx)
+		if err != nil {
+			return err
+		}
```
Comment (Jim3Things) [inline simulation/internal/clients/inventory/inventory.go:758]: Truly a nit, but why not use the combined if form here?  (similar consistency questions elsewhere)

File: simulation/internal/clients/inventory/inventory.go [hunk 2]
```diff
@@ -578,17 +797,200 @@ func (m *Inventory) reconcileNewInventory(
 	m.mutex.Lock()
 	defer m.mutex.Unlock()
 
-	err := m.deleteInventoryDefinitionFromStore(ctx, rootStore)
```
Comment (Jim3Things) [inline simulation/internal/clients/inventory/inventory.go:581]: Ok, this is really a stylistic thing, so take it with the appropriate grain of salt.

I'd consider breaking this up into multiple functions that this function calls - at least one per blade/pdu/tor loop, maybe one for each intermediate loop, and probably one for step1 and one for step2.  Just to keep the length short(er).

File: simulation/internal/clients/inventory/definition.go [hunk 3]
```diff
@@ -631,6 +642,14 @@ func (r *Region) GetDefinitionRegion() *pb.Definition_Region {
 	}
 }
 
+// Equal is used to provide a simple equality check for use to determine
+// if the current region matches the supplied definition. Typically used
+// when looking to see if the record has been changed.
+//
+func (r *Region) Equal(d *pb.Definition_Region) bool {
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition.go:649]: Added all the matching NotXxx() functions and update the calls to !EqualXxxx() to use the NotXxxx() versions.

I assume that's what you meant.

File: simulation/internal/clients/inventory/definition_extended_test.go [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+	err = ts.inventory.writeInventoryDefinitionToStore(ctx, root)
+	require.NoError(err)
+
+	rootReload, err := ts.inventory.readInventoryDefinitionFromStore(ctx)
+	require.NoError(err)
+	require.NotNil(rootReload)
+
+	err = ts.inventory.deleteInventoryDefinitionFromStore(ctx, rootReload)
+	require.NoError(err)
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition_extended_test.go:72]: There is an overall celanup pass invoked as part of SetupSuite() on the base class so no problem there.

This is actually part of the test.  There should really be a verification that at least some select points are there before the delete is invoked and gone after the call. I'll add some.

File: simulation/internal/clients/inventory/inventory.go [hunk 5]
```diff
@@ -552,23 +548,246 @@ func (m *Inventory) DeleteInventoryDefinition(ctx context.Context) error {
...
+	ctx         context.Context,
+	regionStore *Region,
+	regionFile  *pb.Definition_Region,
+) error {
+	errNotFound := errors.ErrRegionNotFound{Region: regionStore.Region}
+
+	_, err := regionStore.Read(ctx)
+
+	if err == errNotFound {
+		tracing.Info(ctx, "   creating new region %s", regionStore.Region)
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go:570]: Done the ones in this file. I'll remove others as I come across them.

File: simulation/internal/clients/inventory/inventory.go (hunk 1 above)
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go:758]: I've cleaned up a bunch of there. In fact that lead to some overall logic simplification in the reconcileXxxXxx() functions. I'm somewhat curious to see if you think that a worthwhile change.

File: simulation/internal/clients/inventory/definition_extended_test.go [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+	ts.inventory = NewInventory(ts.cfg, ts.store)
+}
+
+func (ts *definitionExtendedTestSuite) SetupTest() {
+	ts.testSuiteCore.SetupTest()
+}
+
+func (ts *definitionExtendedTestSuite) TearDownTest() {
+	ts.testSuiteCore.TearDownTest()
+}
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition_extended_test.go:28]: Verified that the calls to the base struct do in fact happen if there is no higher level override.

Removed the redundant calls.

File: simulation/internal/clients/inventory/definition.go [hunk 7]
```diff
@@ -2080,6 +2146,24 @@ func (p *Pdu) GetPorts() *map[int64]*pb.PowerPort {
...
+func (p *Pdu) EqualPorts(ports map[int64]*pb.PowerPort) bool {
+	if len(*p.ports) != len(ports) {
+		return false
+	}
+
+	for i, pp := range *p.ports {
+		if !pp.EqualPort(ports[i]) {
+			return false
+		}
+	}
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition.go:2162]: Agreed. If you can see a way to do this...

File: simulation/internal/clients/inventory/definition_test.go [hunk 8]
```diff
@@ -417,16 +417,11 @@ func (ts *definitionTestSuite) SetupSuite() {
...
-
-	require.NoError(ts.store.Connect())
+	ts.testSuiteCore.SetupTest()
 }
 
 func (ts *definitionTestSuite) TearDownTest() {
-	ts.store.Disconnect()
-	ts.utf.Close()
+	ts.testSuiteCore.TearDownTest()
 }
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition_test.go:425]: As above...

Verified that the calls to the base struct do in fact happen if there is no higher level override.

Removed the redundant calls.

File: simulation/internal/clients/inventory/inventory.go [hunk 9]
```diff
@@ -552,23 +548,246 @@ func (m *Inventory) DeleteInventoryDefinition(ctx context.Context) error {
...
+}
+
+func (m *Inventory) reconcileNewInventoryRegion(
+	ctx         context.Context,
+	regionStore *Region,
+	regionFile  *pb.Definition_Region,
+) error {
+	errNotFound := errors.ErrRegionNotFound{Region: regionStore.Region}
+
+	_, err := regionStore.Read(ctx)
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go:567]: Interesting. I actually have a definite preference for the existing order along the lines of "init all your variables first then do stuff". Probably a result of my long term use of older languages.

I've tried using a cast instead here which removes the need for the variable but does lead to a longer statement so see if you prefer that.

I'd be reluctant to multi-line the longer arg to the cast in the if statement so if that bugs you then I'll go back to the intermediate variable.

File: simulation/internal/clients/inventory/inventory.go [hunk 10]
```diff
@@ -552,23 +548,246 @@ func (m *Inventory) DeleteInventoryDefinition(ctx context.Context) error {
...
+}
+
+func (m *Inventory) reconcileNewInventoryZone(
+	ctx         context.Context,
+	zoneStore   *Zone,
+	zoneFile    *pb.Definition_Zone,
+) error {
+	errNotFound := errors.ErrZoneNotFound{Region: zoneStore.Region, Zone: zoneStore.Zone}
+
+	_, err := zoneStore.Read(ctx)
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go:601]: See above.

File: simulation/internal/clients/inventory/definition.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Jim3Things) [reply simulation/internal/clients/inventory/definition.go:649]: That's great, but not what I meant :)

My comment was very trivial - literally why ".Equal" instead of ".Equals".  I thought the latter read better.

[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition.go:649]: Oh, right, I get it now. 

I'd point out that testify uses Equal() and NotEqual(). In addition, I've found a number of other packages that also provide an Equal() method for equality testing.

As a result, I think these should probably remain as is based on a compatibility argument.

[the remark this answers is not in the document]
Comment (Jim3Things) [reply simulation/internal/clients/inventory/definition.go:649]: Sounds fine. It was a nit, and you've got good reasons here for keeping it this way.

File: simulation/internal/clients/inventory/inventory.go (hunk 2 above)
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go:581]: I've re-arranged some of the code here and moved some of the innermost code loops into sub-routines. The top level functions is definitely shorted and I believe more comprehensible. The cost of course is more functions.

I believe the tradeoff is worth it, thought I'm unsure if this is the *best* arrangement. 

Let me know what you think.

File: simulation/internal/clients/inventory/definition_extended_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition_extended_test.go:64]: Added some verification. The choice of object used for the absence/presence test was an arbitrary choice.

**Comment by MikeCJohnsonUk** [comment]
In addition to response to the PR comments, I've made some changes to the way the equality testing works. Basically, where the underlying object was defined in a proto I've moved the object equality test to a method on the relevant object where possible. I've also updated the tests to use the proto defined GetXxx() methods and/or nil check along the lines of the proto generated methods.

Let me know what you think.

File: simulation/internal/clients/inventory/inventory.go [hunk 11]
```diff
@@ -552,23 +548,360 @@ func (m *Inventory) DeleteInventoryDefinition(ctx context.Context) error {
...
+	torsFile  *map[int64]*pb.Definition_Tor,
+) error {
+	for index := range *torsStore {
+		tor, err := rack.NewTor(index)
+		if err != nil {
+			return err
+		}
+
+		_, ok := (*torsFile)[index]
+		if !ok {
```
Comment (Jim3Things) [inline simulation/internal/clients/inventory/inventory.go:859]: I liked what you did farther down where you combined these into an if statement

**Review by Jim3Things — APPROVED** [review]
Two optional comments that you can either fix or not, now or later.

File: simulation/internal/clients/inventory/definition_test.go [hunk 12]
```diff
@@ -3242,132 +3229,401 @@ func (ts *definitionTestSuite) TestReadInventoryFromStore() {
...
+	err = ts.inventory.UpdateInventoryDefinition(ctx, "./testdata/ChainedBase")
 	require.NoError(err)
 
-	err = ts.inventory.deleteInventoryDefinitionFromStore(ctx, root)
+	// Since nothing changed, we expect no events.
+	//
+	select {
+	case ev, ok := <-watch.Events:
+		require.False(ok)
+		require.Nil(ev)
```
[the remark this answers is not in the document]
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/definition_test.go]: Huh. It would.

Done.

File: simulation/internal/clients/inventory/inventory.go (hunk 11 above)
Comment (MikeCJohnsonUk) [reply simulation/internal/clients/inventory/inventory.go]: Well, in that case!

Actually, it does make sense. Done.

**Comment by MikeCJohnsonUk** [comment]
I fixed those two items and went ahead with the squash and merge. After testing of course...
